hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java

```diff
@@ -1427,21 +1436,136 @@ public void deleteContainer(Container container, boolean force)
+        merkleTree.addChunks(blockData.getLocalID(), chunkInfos);
+      }
+    }
+    checksumManager.writeContainerDataTree(containerData, merkleTree);
+    return merkleTree.toProto().getDataChecksum();
+  }
+
+  private void handleMissingBlock(KeyValueContainer container, ContainerData containerData, TokenHelper tokenHelper,
+                                  long scmBlockSize, XceiverClientSpi xceiverClient,
+                                  ContainerProtos.BlockMerkleTree missingBlock) throws IOException {
+    BlockID blockID = new BlockID(containerData.getContainerID(), missingBlock.getBlockID());
+    Token<OzoneBlockTokenIdentifier> blockToken = tokenHelper.getBlockToken(blockID, scmBlockSize);
+    // TODO: Cache the blockResponse to reuse it again.
+    ContainerProtos.GetBlockResponseProto blockResponse = ContainerProtocolCalls.getBlock(xceiverClient, blockID,
+        blockToken, new HashMap<>());
+    // TODO: Add BcsId in BlockMerkleTree to avoid this call
+    ContainerProtos.GetCommittedBlockLengthResponseProto blockLengthResponse =
+        ContainerProtocolCalls.getCommittedBlockLength(xceiverClient, blockID, blockToken);
+    List<ContainerProtos.ChunkInfo> chunksList = blockResponse.getBlockData().getChunksList();
+
+    for (ContainerProtos.ChunkInfo chunkInfoProto : chunksList) {
+      ByteString chunkData = readChunkData(xceiverClient, chunkInfoProto, blockID, blockToken);
+      ChunkBuffer chunkBuffer = ChunkBuffer.wrap(chunkData.asReadOnlyByteBuffer());
+      ChunkInfo chunkInfo = ChunkInfo.getFromProtoBuf(chunkInfoProto);
+      chunkInfo.addMetadata(OzoneConsts.CHUNK_OVERWRITE, "true");
+      writeChunkForClosedContainer(chunkInfo, blockID, chunkBuffer, container);
+    }
+
+    putBlockForClosedContainer(chunksList, container, BlockData.getFromProtoBuf(blockResponse.getBlockData()),
+        blockLengthResponse.getBlockLength());
+  }
+
+  private ByteString readChunkData(XceiverClientSpi xceiverClient, ContainerProtos.ChunkInfo chunkInfoProto,
+                                   BlockID blockID, Token<OzoneBlockTokenIdentifier> blockToken) throws IOException {
+    ContainerProtos.ReadChunkResponseProto response =
+        ContainerProtocolCalls.readChunk(xceiverClient, chunkInfoProto, blockID.getDatanodeBlockIDProtobuf(),
+            null, blockToken);
+
+    if (response.hasData()) {
+      return response.getData();
+    } else if (response.hasDataBuffers()) {
+      return BufferUtils.concatByteStrings(response.getDataBuffers().getBuffersList());
+    } else {
+      throw new IOException("Error reading chunk data: No data returned.");
+    }
+  }
+
+  private void reconcileChunk(KeyValueContainer container, ContainerData containerData, TokenHelper tokenHelper,
+                              long scmBlockSize, XceiverClientSpi xceiverClient,
+                              Map.Entry<Long, List<ContainerProtos.ChunkMerkleTree>> mapEntry) throws IOException {
```

Comment:
It will be cleaner to split the map entry into two separate parameters. This method does not need to know that the info came from map in the caller.